### PR TITLE
ref(metric alerts): Create new consumer using arroyo

### PR DESCRIPTION
**Pull request by ceorourke**
High metric alert volume is causing notification delays. This PR implements (but does not yet use) a new consumer that uses arroyo so we can take advantage of multi threaded processing. 

See [WOR-2747](https://getsentry.atlassian.net/browse/WOR-2747)

[WOR-2747]: https://getsentry.atlassian.net/browse/WOR-2747?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/sentry/snuba/query_subscription_consumer.py [hunk 1]
```diff
@@ -46,22 +212,89 @@ class InvalidSchemaError(InvalidMessageError):
...
+        cluster_name,
+        {
+            "group.id": group_id,
+            "session.timeout.ms": 6000,
+            "auto.offset.reset": initial_offset_reset,
+            "enable.auto.commit": "false",
+            "enable.auto.offset.store": "false",
+            "enable.partition.eof": "false",
+            "default.topic.config": {"auto.offset.reset": initial_offset_reset},
+        },
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:273]:
```suggestion
```

You should be able to remove these lines, the `build_kafka_consumer_configuration` function should build the config with the guarantees and configuration Arroyo expects.

File: src/sentry/snuba/query_subscription_consumer.py [hunk 2]
```diff
@@ -38,6 +59,151 @@ def inner(func: TQuerySubscriptionCallable) -> TQuerySubscriptionCallable:
...
+    matches the expected schema.
+    :param value: A json formatted string
+    :return: A dict with the parsed message
+    """
+    with metrics.timer("snuba_query_subscriber.parse_message_value.json_parse"):
+        wrapper: Dict[str, Any] = json.loads(value)
+
+    with metrics.timer("snuba_query_subscriber.parse_message_value.json_validate_wrapper"):
+        try:
+            jsonschema.validate(wrapper, SUBSCRIPTION_WRAPPER_SCHEMA)
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:74]: Not for now as there's enough changes happening here but later on you could use https://github.com/getsentry/arroyo/blob/56066e373bd196482cd40ded8a5a613a7486a202/arroyo/processing/strategies/decoder/json.py in Arroyo for this. It will probably be faster for no other reason but that it uses fastjsonschema.

File: src/sentry/snuba/query_subscription_consumer.py [hunk 3]
```diff
@@ -46,22 +216,81 @@ class InvalidSchemaError(InvalidMessageError):
...
+        commit: Commit,
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        def process_message(message: Message[KafkaPayload]) -> None:
+            with sentry_sdk.start_transaction(
+                op="handle_message",
+                name="query_subscription_consumer_process_message",
+                sampled=random() <= options.get("subscriptions-query.sample-rate"),
+            ), metrics.timer("snuba_query_subscriber.handle_message"):
+                value: Any = message.value
```
Comment (ceorourke) [inline src/sentry/snuba/query_subscription_consumer.py:234]: I probably need to fix this up, was getting irritated with typing.

File: src/sentry/snuba/query_subscription_consumer.py [hunk 4]
```diff
@@ -46,22 +216,81 @@ class InvalidSchemaError(InvalidMessageError):
...
+    cluster_options = kafka_config.get_kafka_consumer_cluster_options(cluster_name)
+    consumer = KafkaConsumer(
+        build_kafka_consumer_configuration(
+            cluster_options,
+            auto_offset_reset=auto_offset_reset,
+            group_id=group_id,
+            strict_offset_reset=strict_offset_reset,
+        )
+    )
+    return StreamProcessor(
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:279]: If you want to configure automatic metrics via Arroyo for this consumer you need to call `configure_metrics` with your metrics backend before this line. There is no docs page for this yet, but I wrote up some information about it here: https://github.com/getsentry/arroyo/pull/198

File: src/sentry/snuba/query_subscription_consumer.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/snuba/query_subscription_consumer.py:234]: Ah yeah, this is a bit confusing since there are 2 types of message values in the library. 

For now you can do something like:
```
value = message.value
assert isinstance(value, BrokerValue)
offset = value.offset
partition = value.partition.index
```

File: src/sentry/runner/commands/run.py [hunk 5]
```diff
@@ -427,20 +427,32 @@ def post_process_forwarder(**options):
     type=click.Choice(["earliest", "latest"]),
     help="Force subscriptions to start from a particular offset",
 )
+@click.option(
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply src/sentry/runner/commands/run.py:430]: Guessing you want to test this in SaaS on one of the subscription types to start? Basically, make an ops request to pass it. Once you feel confident everything is working, flip the default to True to roll it out everywhere.

File: src/sentry/runner/commands/run.py [hunk 6]
```diff
@@ -427,20 +427,35 @@ def post_process_forwarder(**options):
...
+    "--use-arroyo-consumer",
+    default=False,
+    is_flag=True,
+    help="Switches to the new arroyo consumer implementation.",
+)
```
Comment (lynnagara) [inline src/sentry/runner/commands/run.py:435]: Arroyo consumers handle auto offset reset a bit differently, mind adding the `@strict_offset_reset_option()` decorator here and passing it's value through to the `KafkaConsumer`

File: src/sentry/snuba/query_subscription_consumer.py [hunk 7]
```diff
@@ -46,22 +217,77 @@ class InvalidSchemaError(InvalidMessageError):
...
+                name="query_subscription_consumer_process_message",
+                sampled=random() <= options.get("subscriptions-query.sample-rate"),
+            ), metrics.timer("snuba_query_subscriber.handle_message"):
+                value = message.value
+                assert isinstance(value, BrokerValue)
+                offset = value.offset
+                partition = value.partition.index
+                try:
+                    handle_message(
+                        str(message.payload.value),
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:241]: Why did you need to cast the bytes to a string?

File: src/sentry/runner/commands/run.py [hunk 8]
```diff
@@ -427,20 +427,35 @@ def post_process_forwarder(**options):
...
+        subscriber = QuerySubscriptionConsumer(
+            group_id=options["group"],
+            topic=options["topic"],
+            commit_batch_size=options["commit_batch_size"],
+            commit_batch_timeout_ms=options["commit_batch_timeout_ms"],
+            initial_offset_reset=options["initial_offset_reset"],
+            force_offset_reset=options["force_offset_reset"],
+        )
+    else:
+        subscriber = get_query_subscription_consumer(
```
Comment (lynnagara) [inline src/sentry/runner/commands/run.py:454]: Pass the 2 offset reset options (--auto-offset-reset and --no-strict-offset-reset) into here too from the CLI

File: src/sentry/snuba/query_subscription_consumer.py [hunk 9]
```diff
@@ -38,6 +62,153 @@ def inner(func: TQuerySubscriptionCallable) -> TQuerySubscriptionCallable:
     return inner
 
 
+def parse_message_value(value: str) -> Dict[str, Any]:
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:65]: I didn't read through this diff, assuming it's mostly just moved code

File: src/sentry/snuba/query_subscription_consumer.py (hunk 7 above)
Comment (ceorourke) [reply src/sentry/snuba/query_subscription_consumer.py]: Hmm at the time I was getting a JSON decode error but it seems to run fine w/o it now, updated.

File: src/sentry/snuba/query_subscription_consumer.py (hunk 9 above)
Comment (ceorourke) [reply src/sentry/snuba/query_subscription_consumer.py:65]: Yes this function is untouched, just moved.

File: src/sentry/runner/commands/run.py (hunk 8 above)
Comment (ceorourke) [reply src/sentry/runner/commands/run.py:455]: Where do I use them?

**Comment by ceorourke** [comment]
Btw I am still doing manual local testing - I had some dev env issues today that slowed me down and now things don't error but also don't create a metric alert lol ... I'll keep looking into it tomorrow,

**Comment by lynnagara** [comment]
@ceorourke Where does topic `f187d2542a87419599159f4e987bd22d` come from? i don't recognise that at all.

File: src/sentry/runner/commands/run.py (hunk 8 above)
Comment (lynnagara) [reply src/sentry/runner/commands/run.py:455]: Pass to KafkaConsumer. Example of how this: https://github.com/getsentry/sentry/blob/b134ebb0d4cbe3d5a03cd633d4bbb1c29cbd8545/src/sentry/eventstream/kafka/backend.py#L245-L252

File: src/sentry/runner/commands/run.py [hunk 10]
```diff
@@ -427,20 +427,39 @@ def post_process_forwarder(**options):
...
+            topic=options["topic"],
+            group_id=options["group"],
+            strict_offset_reset=options["strict_offset_reset"],
+            initial_offset_reset=options["initial_offset_reset"],
+            force_offset_reset=options["force_offset_reset"],
```
Comment (lynnagara) [inline src/sentry/runner/commands/run.py:460]: Don't pass the `force_offset_reset` in here. We're trying to deprecate that from the codebase entirely and just use the other 2 offset reset parameters everywhere.

File: tests/snuba/incidents/test_tasks.py [hunk 11]
```diff
@@ -158,3 +160,16 @@ def shutdown_callback(*args, **kwargs):
...
+
+    def test(self):
+        consumer = QuerySubscriptionConsumer("hi", topic=self.topic)
+        self.run_test(consumer)
+
+    def test_arroyo(self):
+        from sentry.utils.batching_kafka_consumer import create_topics
+
+        kafka_cluster = settings.KAFKA_TOPICS[self.topic]["cluster"]
+        create_topics(kafka_cluster, [self.topic])
```
Comment (ceorourke) [inline tests/snuba/incidents/test_tasks.py:172]: I couldn't get your suggested [create_topic](https://github.com/getsentry/sentry/blob/master/tests/sentry/eventstream/kafka/test_consumer.py#L23-L45) working, it always threw an error on the creation command, but this works fine!

```
________________________________________________________________ HandleSnubaQueryUpdateTest.test_arroyo _________________________________________________________________
tests/snuba/incidents/test_tasks.py:203: in test_arroyo
    with self.create_topic(self.topic):
../../.pyenv/versions/3.8.13/lib/python3.8/contextlib.py:113: in __enter__
    return next(self.gen)
tests/snuba/incidents/test_tasks.py:70: in create_topic
    subprocess.check_call(
../../.pyenv/versions/3.8.13/lib/python3.8/subprocess.py:364: in check_call
    raise CalledProcessError(retcode, cmd)
E   subprocess.CalledProcessError: Command '['docker', 'exec', 'sentry_kafka', 'kafka-topics', '--zookeeper', '127.0.0.1:2181', '--create', '--topic', '4750ba1012d74b729c06585f989ed7d1', '--partitions', '1', '--replication-factor', '1']' returned non-zero exit status 1.
------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------
Exception in thread "main" kafka.zookeeper.ZooKeeperClientTimeoutException: Timed out waiting for connection while in state: CONNECTING
	at kafka.zookeeper.ZooKeeperClient.waitUntilConnected(ZooKeeperClient.scala:271)
	at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:125)
	at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1948)
	at kafka.admin.TopicCommand$ZookeeperTopicService$.apply(TopicCommand.scala:378)
	at kafka.admin.TopicCommand$.main(TopicCommand.scala:56)
	at kafka.admin.TopicCommand.main(TopicCommand.scala)
```

Comment (lynnagara) [reply tests/snuba/incidents/test_tasks.py:172]: Ok. Yeah I'm not a fan of the subprocess.call either, better using the client API to create the topic which is what `create_topics` does. Ideally we would also delete the topic after the test though

File: src/sentry/snuba/query_subscription_consumer.py [hunk 12]
```diff
@@ -46,22 +217,82 @@ class InvalidSchemaError(InvalidMessageError):
...
+                    )
+
+        return RunTask(process_message, CommitOffsets(commit))
+
+
+def get_query_subscription_consumer(
+    topic: str,
+    group_id: str,
+    strict_offset_reset: bool,
+    initial_offset_reset: str = "earliest",
```
Comment (lynnagara) [inline src/sentry/snuba/query_subscription_consumer.py:267]: Please don't set this default here, force the caller to pass it. It's very unexpected that it's different from the default in the CLI which is latest.

Comment (ceorourke) [reply src/sentry/snuba/query_subscription_consumer.py:267]: Oops will do, I was copying what the existing consumer had.

File: tests/sentry/eventstream/kafka/test_consumer.py [hunk 13]
```diff
@@ -156,4 +158,4 @@ def test_post_process_forwarder_streaming_consumer(self, dispatch_post_process_g
             occurrence_id=None,
         )
 
-        consumer.signal_shutdown()
+        consumer._shutdown()
```
Comment (ceorourke) [inline tests/sentry/eventstream/kafka/test_consumer.py:161]: I wasn't able to address the test failure here w/o changing this - it always hung and gave an error about the `sentry_test` database already existing. It seems like `signal_shutdown` is meant for `run` but we used `run_once` so `_shutdown` is appropriate.

**Comment by ceorourke** [comment]
I left a comment with follow up tasks here: https://getsentry.atlassian.net/browse/WOR-2747?focusedCommentId=67229

**Review by lynnagara — APPROVED** [review]
👯
